crates/tempfile/src/lib.rs

```diff
@@ -41,24 +48,64 @@ impl Drop for TempDir {
     }
 }
 
-fn get_template() -> PathBuf {
-    std::env::temp_dir().join("fish_tmp_XXXXXX")
+// Creates a random path in the directory `dir`.
+// The last component of the path will start with `prefix`, followed by a random sequence of
+// alphanumeric ASCII characters.
+pub fn get_random_path_in_dir(dir: PathBuf, mut prefix: OsString) -> PathBuf {
+    let mut rng = rand::rng();
+    let mut random_char_iter = Alphanumeric.sample_iter(&mut rng);
+    let mut random_part = String::new();
+    for _ in 0..10 {
+        random_part.push(random_char_iter.next().unwrap().into());
+    }
+    prefix.push(random_part);
+    dir.join(prefix)
+}
+
+/// Tries to create a new file at the path returned by `generate_path`.
+/// If a file already exists at this path, `generate_path` will be called again and file creation
+/// will be retried. This is repeated until a new file is created, or an error occurs.
+pub fn create_file_with_retry<F: Fn() -> PathBuf>(
```

Comment:
yeah this seems like an appropriate API for our use cases
(since we use both std::env::temp_dir and ~/.local/share/fish as TMPDIR).